cpp/src/arrow/flight/sql/odbc/odbc_api.cc

```diff
@@ -1308,4 +1308,115 @@ SQLRETURN SQLNativeSql(SQLHDBC connectionHandle, SQLWCHAR* inStatementText,
   });
 }
 
+SQLRETURN SQLDescribeCol(SQLHSTMT stmt, SQLUSMALLINT columnNumber, SQLWCHAR* columnName,
+                         SQLSMALLINT bufferLength, SQLSMALLINT* nameLengthPtr,
+                         SQLSMALLINT* dataTypePtr, SQLULEN* columnSizePtr,
+                         SQLSMALLINT* decimalDigitsPtr, SQLSMALLINT* nullablePtr) {
+  LOG_DEBUG(
+      "SQLDescribeColW called with stmt: {}, columnNumber: {}, "
+      "columnName: {}, bufferLength: {}, nameLengthPtr: {}, dataTypePtr: {}, "
+      "columnSizePtr: {}, decimalDigitsPtr: {}, nullablePtr: {}",
+      stmt, columnNumber, fmt::ptr(columnName), bufferLength, fmt::ptr(nameLengthPtr),
+      fmt::ptr(dataTypePtr), fmt::ptr(columnSizePtr), fmt::ptr(decimalDigitsPtr),
+      fmt::ptr(nullablePtr));
+  using ODBC::ODBCDescriptor;
+  using ODBC::ODBCStatement;
+
+  return ODBCStatement::ExecuteWithDiagnostics(stmt, SQL_ERROR, [=]() {
+    ODBCStatement* statement = reinterpret_cast<ODBCStatement*>(stmt);
+    ODBCDescriptor* ird = statement->GetIRD();
+
+    SQLSMALLINT sqlType;
+    ird->GetField(columnNumber, SQL_DESC_CONCISE_TYPE, &sqlType, sizeof(SQLSMALLINT),
+                  nullptr);
+
+    SQLINTEGER outputLengthInt;
+    ird->GetField(columnNumber, SQL_DESC_NAME, columnName, bufferLength,
+                  &outputLengthInt);
+    if (nameLengthPtr) {
+      *nameLengthPtr = static_cast<SQLSMALLINT>(outputLengthInt);
+    }
+
+    if (dataTypePtr) {
+      *dataTypePtr = sqlType;
+    }
+
+    // Column Size
+    if (columnSizePtr) {
+      switch (sqlType) {
+        // All numeric types
+        case SQL_DECIMAL:
+        case SQL_NUMERIC:
+        case SQL_TINYINT:
+        case SQL_SMALLINT:
+        case SQL_INTEGER:
+        case SQL_BIGINT:
+        case SQL_REAL:
+        case SQL_FLOAT:
+        case SQL_DOUBLE: {
+          ird->GetField(columnNumber, SQL_DESC_PRECISION, columnSizePtr, sizeof(SQLULEN),
+                        nullptr);
+          break;
+        }
+
+        default: {
+          ird->GetField(columnNumber, SQL_DESC_LENGTH, columnSizePtr, sizeof(SQLULEN),
+                        nullptr);
+        }
+      }
+    }
+
+    // Decimal Digits
+    if (decimalDigitsPtr) {
+      switch (sqlType) {
+        // All exact numeric types
+        case SQL_TINYINT:
+        case SQL_SMALLINT:
+        case SQL_INTEGER:
+        case SQL_BIGINT:
+        case SQL_DECIMAL:
+        case SQL_NUMERIC: {
+          ird->GetField(columnNumber, SQL_DESC_SCALE, decimalDigitsPtr, sizeof(SQLULEN),
+                        nullptr);
+          break;
+        }
+
+        // All datetime types
+        case SQL_DATE:
+        case SQL_TIME:
+        case SQL_TIMESTAMP:
```

Comment:
```suggestion
        // All datetime types
        case SQL_DATE:
        case SQL_TIME:
        case SQL_TIMESTAMP:
        case SQL_TYPE_DATE:
        case SQL_TYPE_TIME:
        case SQL_TYPE_TIMESTAMP:
```
we should add `SQL_TYPE_*` for ODBC 3 support since they map to different values; feel free to reorder the types and regroup the same data type for ODBC 2 and 3 together